src/controllers/controllerengine.cpp

```diff
@@ -838,46 +821,55 @@ QScriptValue ControllerEngine::connectControl(QString group, QString name,
             proxy->disconnect();
         }
     } else {
-        qWarning() << "Invalid callback";
+        qWarning() << "Cannot make connection to" << group << name
+                   << ": no such callback function";
         return QScriptValue(false);
     }
 
-    if (function.isFunction()) {
-        qDebug() << "Connection:" << group << name;
+    QScriptContext *ctxt = m_pEngine->currentContext();
+    // Our current context is a function call to engine.connectControl. We
+    // want to grab the 'this' from the caller's context, so we walk up the
+    // stack.
+    if (ctxt) {
+        ctxt = ctxt->parentContext();
+        conn.context = ctxt ? ctxt->thisObject() : QScriptValue();
+    }
+    
+    if (!callback.isString()) {
+        QUuid uuid = QUuid::createUuid();
+        conn.id = uuid.toString();
+    }
+    
+    if (disconnect) {
+        return QScriptValue(disconnectControl(conn));
+    } else {
+        if (m_connectedControls.contains(conn.key, conn)) {
+            qWarning() << "Cannot connect" << conn.key << "to"
+                        << conn.id << ": already connected";
+            return QScriptValue(false);
+        }
+        
         connect(cot, SIGNAL(valueChanged(double)),
                 this, SLOT(slotValueChanged(double)),
                 Qt::QueuedConnection);
         connect(cot, SIGNAL(valueChangedByThis(double)),
                 this, SLOT(slotValueChanged(double)),
                 Qt::QueuedConnection);
+        m_connectedControls.insert(conn.key, conn);
 
-        ControllerEngineConnection conn;
-        conn.key = key;
-        conn.ce = this;
-        conn.function = function;
-
-        QScriptContext *ctxt = m_pEngine->currentContext();
-        // Our current context is a function call to engine.connectControl. We
-        // want to grab the 'this' from the caller's context, so we walk up the
-        // stack.
-        if (ctxt) {
-            ctxt = ctxt->parentContext();
-            conn.context = ctxt ? ctxt->thisObject() : QScriptValue();
-        }
+        if (m_connectedControls.contains(conn.key, conn)) {
+            qDebug() << "Connection: function" << conn.id
+                     << "connected to" << conn.key;
+            // Return a wrapper to the conn
+            QHash<ConfigKey, ControllerEngineConnection>::iterator i =
+                m_connectedControls.find(conn.key);
```

Comment:
I do not understand the call, since this is a multi hash. Do we get the fist connection to the key?